rfcs/rfc-001-smart-contracts-engine.md

```diff
@@ -37,154 +37,175 @@ version 0.1
-
-### State and Entry Point
-
-State is addressed by an account which is at the moment simply the PubKey.  Our goal is to eliminate dynamic memory allocation in the smart contract itself, so the contract is a function that takes a mapping of [(PubKey,State)] and returns [(PubKey, State')].  The output of keys is a subset of the input.  Three basic kinds of state exist:
+The goal with the runtime is to have a general purpose execution environment that is highly parallelizeable and doesn't require dynamic resource management. The goal is to execute as many contracts as possible in parallel, and have them pass or fail without a destructive state change.
 
-* Instance State
-* Participant State
-* Caller State
 
-There isn't any difference in how each is implemented, but conceptually Participant State is memory that is allocated for each participant in the contract.  Instance State is memory that is allocated for the contract itself, and Caller State is memory that the transactions caller has allocated.
+### State
 
+State is addressed by an account which is at the moment simply the Pubkey.  Our goal is to eliminate memory allocation from within the smart contract itself.  Thus the client of the contract provides all the state that is necessary for the contract to execute in the transaction itself.  The runtime interacts with the contract through a state transition function, which takes a mapping of [(Pubkey,State)] and returns [(Pubkey, State')].  The State is an opeque type to the runtime, a `Vec<u8>`, the contents of which the contract has full control over.
 
-### Call
-
+### Call Structure
 ```
-void call(
-    const struct instance_data *data,
-    const uint8_t kind[],  //instance|participant|caller|read|write
-    const uint8_t *keys[],
-    uint8_t *data[],
-    int num,
-    uint8_t dirty[],        //dirty memory bits
-    uint8_t *userdata,      //current transaction data
-);
+/// Call definition
+/// Signed portion
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub struct CallData {
+    /// Each Pubkey in this vector is mapped to a corresponding `Page` that is loaded for contract execution
+    /// In a simple pay transaction `key[0]` is the token owner's key and `key[1]` is the recipient's key.
+    pub keys: Vec<Pubkey>,
+
+    /// The Pubkeys that are required to have a proof.  The proofs are a `Vec<Signature> which encoded along side this data structure
+    /// Each Signature signs the `required_proofs` vector as well as the `keys` vectors.  The transaction is valid if and only if all
+    /// the required signatures are present and the public key vector is unchanged between signatures.
+    pub required_proofs: Vec<u8>,
+
+    /// PoH data
+    /// last PoH hash observed by the sender
+    pub last_id: Hash,
+
+    /// Program
+    /// The address of the program we want to call.  ContractId is just a Pubkey that is the address of the loaded code that will execute this Call.
+    pub contract_id: ContractId,
```

Comment:
just a naming thing before we write too much code:  ContractId/contract_id or ProgramId/program_id?   I prefer the latter.   We execute programs, some of which may be contracts.  But ultimately I'm just looking for consistency.

Reply:
@mvines they call all the on chain stuff contracts.  not sure why :)

Reply:
That's good right, we can say we're different cause we run programs and not only contracts

Reply:
The term contract is already in use within this codebase. If we're going to change that, I'd prefer to see it in a separate PR that changes all those uses at once.

Also, it's not obvious to me that "program" is the most precise term. If we're going to change it, we should give it a bunch of thought and hopefully only need to change it once.